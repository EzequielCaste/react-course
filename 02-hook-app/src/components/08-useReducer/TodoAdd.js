import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAddTodo }) => {

  const [ { description }, handleInputChange, reset ] = useForm({
    description: ''
  }); 

  const handleSubmit = (e) => {
    e.preventDefault();

    if ( description.trim().length <= 1 ) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };
  
    handleAddTodo( newTodo );
    reset();
  }

  return (
    <>
      <h4>Agregar Tarea</h4>
      <hr/>

      <form onSubmit={ handleSubmit }>
        <input 
          value={ description }
          onChange={ handleInputChange }
          type="text"
          name="description"
          placeholder="Tarea..."
          autoComplete="false"
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block">
          Agregar
        </button>
      </form>        
    </>
  )
}
