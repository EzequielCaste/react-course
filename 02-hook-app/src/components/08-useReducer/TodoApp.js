import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm'

import './styles.css'

const init = () => {
  
  return JSON.parse(localStorage.getItem('todos')) || [];
 
}

export const TodoApp = () => {

  const [ todos, dispatch ] = useReducer(todoReducer, [], init);

  const [ { description }, handleInputChange, reset ] = useForm({
    description: ''
  }); 

  useEffect( () => {
    localStorage.setItem('todos', JSON.stringify( todos ))
  }, [todos]);

  const handleDelete = ( todoId ) => {
    
    const action = {
      type: 'delete',
      payload: todoId
    }

    dispatch( action )
  }

  const handleSubmit = (e) => {
    e.preventDefault();

    if ( description.trim().length <= 1 ) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    const action = {
      type: 'add',
      payload: newTodo
    }

    dispatch( action );
    reset();


  }
  return (
    <div>
      <h1>TodoApp ( {todos.length} )</h1>
      <hr/>
      <div className="row">
        <div className="col-7">
        <ul className="list-group list-group-flush">
            {
              todos.map( (todo, i) => (
                <li
                  className="list-group-item"
                  key={ todo.id }
                >
                <p className="text-center">            
                { i + 1 } { todo.desc } 
                </p>    
                <button
                  onClick={ () => handleDelete(todo.id) }
                  className="btn btn-danger"
                >
                  Borrar
                </button>
                </li>
                )
              )
            }
          </ul>
        </div>
        <div className="col-5">
            <h4>Agregar Tarea</h4>
            <hr/>

            <form onSubmit={ handleSubmit }>
              <input 
                value={ description }
                onChange={ handleInputChange }
                type="text"
                name="description"
                placeholder="Tarea..."
                autoComplete="false"
              />
              <button
               type="submit"
               className="btn btn-outline-primary mt-1 btn-block">
                Agregar
              </button>
            </form>

        </div>
      </div>
    </div>
  )
}
