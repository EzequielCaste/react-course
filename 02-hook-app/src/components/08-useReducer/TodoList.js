import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map( (todo, i) => (
          // TodoListItem todo indice handlers
          <TodoListItem 
            todo={todo} 
            key={todo.id}
            index={i}
            handleToggle={handleToggle}
            handleDelete={handleDelete}
          />
          
          )
        )
      }
    </ul>
  )
}
