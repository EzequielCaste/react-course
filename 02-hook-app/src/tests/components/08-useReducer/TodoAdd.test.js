import React from 'react'
import { shallow } from 'enzyme'
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';
import '@testing-library/jest-dom'

describe('Pruebas en <TodoAdd />', () => {

  const handleAddTodo = jest.fn();

  const wrapper = shallow( 
  <TodoAdd 
    handleAddTodo={ handleAddTodo }
  />);

  test('debe mostrarse correctamente', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('NO debe llamar HandleAddTodo', () => {

    const formSubmit = wrapper.find('form').prop('onSubmit');

    formSubmit({ preventDefault(){} });

    expect( handleAddTodo ).toHaveBeenCalledTimes(0);
  });

  test('debe llamar la funciÃ³n HandleAddTodo', () => {
    // con un argumento // called
    // simular manipulacion en el input
    const value = 'Aprender React';

    wrapper.find('input').simulate('change', {
      target: {
        name: 'description',
        value
      }
    });

   const formSubmit = wrapper.find('form').prop('onSubmit');
   formSubmit({ preventDefault(){} });

   expect( handleAddTodo ).toHaveBeenCalledTimes(1);
   expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) );
   expect( handleAddTodo ).toHaveBeenCalledWith( {
     id: expect.any(Number),
     desc: value,
     done: false
   } );

   expect( wrapper.find('input').prop('value')).toBe('');

  });
})