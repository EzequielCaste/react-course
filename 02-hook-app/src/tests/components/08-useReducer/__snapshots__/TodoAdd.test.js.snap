// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> debe mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar Tarea
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="false"
      name="description"
      onChange={[Function]}
      placeholder="Tarea..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
